packages/react-reconciler/src/ReactFiberHotReloading.js

```diff
@@ -0,0 +1,312 @@
+    if (family === undefined) {
+      // Check if we're dealing with a real forwardRef. Don't want to crash early.
+      if (
+        type !== null &&
+        type !== undefined &&
+        typeof type.render === 'function'
+      ) {
+        // ForwardRef is special because its resolved .type is an object,
+        // but it's possible that we only have its inner render function in the map.
+        // If that inner render function is different, we'll build a new forwardRef type.
+        const currentRender = resolveFunctionForHotReloading(type.render);
+        if (type.render !== currentRender) {
+          return {...type, render: currentRender};
+        }
+      }
+      return type;
+    }
+    // Use the latest known implementation.
+    return family.currentType;
+  } else {
+    return type;
+  }
+}
+
+export function isCompatibleFamilyForHotReloading(
+  fiber: Fiber,
+  element: ReactElement,
+): boolean {
+  if (__DEV__) {
+    if (familiesByType === null) {
+      // Hot reloading is disabled.
+      return false;
+    }
+
+    const prevType = fiber.elementType;
+    const nextType = element.type;
+    // If we got here, we know types aren't === equal.
+    let needsCompareFamilies = false;
+    switch (fiber.tag) {
+      case FunctionComponent: {
+        if (typeof nextType === 'function') {
+          needsCompareFamilies = true;
+        }
+        break;
+      }
+      case ForwardRef: {
+        if (
+          typeof nextType === 'object' &&
+          nextType !== null &&
+          nextType.$$typeof === REACT_FORWARD_REF_TYPE
+        ) {
+          needsCompareFamilies = true;
+        }
+        break;
+      }
+      case MemoComponent:
+      case SimpleMemoComponent: {
+        if (
+          typeof nextType === 'object' &&
+          nextType !== null &&
+          nextType.$$typeof === REACT_MEMO_TYPE
+        ) {
+          // TODO: if it was but can no longer be simple,
+          // we shouldn't set this.
+          needsCompareFamilies = true;
+        }
+        break;
+      }
+      // TODO: maybe support lazy?
+      default:
+        return false;
+    }
+
+    // Check if both types have a family and it's the same one.
+    if (needsCompareFamilies) {
+      // Note: memo() and forwardRef() we'll compare outer rather than inner type.
+      // This means both of them need to be registered to preserve state.
+      // If we unwrapped and compared the inner types for wrappers instead,
+      // then we would risk falsely saying two separate memo(Foo)
+      // calls are equivalent because they wrap the same Foo function.
+      const prevFamily = familiesByType.get(prevType);
+      if (
+        prevFamily !== undefined &&
+        prevFamily === familiesByType.get(nextType)
+      ) {
+        return true;
+      }
+    }
+    return false;
+  } else {
+    return false;
+  }
+}
+
+export function shouldForceRenderForHotReloading(fiber: Fiber | null): boolean {
```

Comment:
Forcing a rerender is a very loosely defined concept. We need to define this more clearly and we shouldn't call it "force" because it doesn't mean anything. E.g. if the rerender returns the same thing, does that mean we can bail out? What if we can determine early that nothing has changed based on a compiler driven memoization in the component? When are we allowed to bail out?

I think this shouldn't actually be treated as a forced rerender. It should be modeled as changing the "type".

`current.type !== workInProgress.type`

We already do this for `current.ref !== workInProgress.ref` and `current.props !== workInProgress.props`. This is just the same thing for a third field. The rest should follow.

What breaks down if you model it that way?

Reply:
The thing that I want to happen is for it to not skip `reconcileFibers`. Then it'll "discover" its child type is now `'DELETED'`, and force it to remount.

Reply:
Why does is triggering `reconcileFibers` in the parent a requirement? In case the tag needs to change?

Reply:
The approach I would suggest off the top of my head (haven't thought through this fully so sorry if there's something obviously wrong)

- Schedule on update the reloaded fiber. Don't do anything special to the parent.
- When you get to `createWorkInProgress`, instead of cloning from `current`, create a new work-in-progress fiber that's disconnected (no alternate).
- For each of the old fiber's children, schedule a deletion effect.

Reply:
Something's gotta tag that child as receiving a deletion? I thought that's what `reconcileChildFibers` does. I'm fuzzy on how child reconciliation works.

Reply:
I need the ability to remount a child in production for dehydrated suspense boundaries. The upgrade logic has a bug otherwise. So we need that anyway but not implemented like this.

We should think of a different technique than rerendering the parent to do this.

Reply:
Also, I need to insert the new retried child at its place.

Reply:
lol github comments are so out of order

Reply:
OK, I can think more about this

Reply:
Btw I'll also need to remount all currently failed boundaries on hot update. Planned to keep them in a weak set. This is so that you can recover after simple errors without refresh.

Reply:
Can you explain the scenario when you need this? We’re trying to guess over here and not succeeding.

Reply:
The scenario where I need what exactly?

Reply:
I have two reasons to want a "force remount" functionality:

* Changing Hooks signature. If some function gets a different set of Hooks, I want to force it to remount without even trying a hot render. However, I don't have a reliable way to propagate new function identity to all callers. So I want to tell React "remount all things with this family during this hot update".

* Failed error boundaries. If my hot edit led to a runtime error (as it often would due to typos etc), closest error boundary will activate together with global runtime error overlay. This is good. But once I save again, I want to "retry". That component doesn't exist anymore, and there's no way to tell error boundary to "reset state" (reset to what?) But I can remount it. That's what I want to do for failed boundaries during a hot update.

Reply:
@gaearon It works by scheduling a `Deletion` effect on the fiber that's deleted and adding them to the parent's effect list. You can do that manually without reconciling the parent.

Reply:
I’d also need to schedule an insertion for the new version, right?

Reply:
An error boundary doesn’t always setState on itself. Sometimes it sets state on a parent. How do you plan to deal with that?

Reply:
> Changing Hooks signature

I think there's some subtle nuances to think about here. I originally thought of this as the "type" changing the family but I think that it explicitly can't use that mechanism. It must preserve the same family because some parent component can still have a React element that references the old function. Rendering a new instance using that element should get the new implementation.

Therefore, what you're looking for here is explicitly to reset the state rendered by the component but the "self" identity from the perspective of the outside conceptually still the same. In fact, you could argue that it shouldn't even delete the old fiber, but remount it in place (although granted, deleting and remounting is might still be the easiest way to do that).

Reply:
>An error boundary doesn’t always setState on itself. Sometimes it sets state on a parent. How do you plan to deal with that?

That's a good point. But I don't have a clear sense of how common that is in practice — especially with the future move to `getDerivedStateFromError`. In that world it seems like it's still a good enough heuristic to reset just the boundary alone.

>It must preserve the same family because some parent component can still have a React element that references the old function. Rendering a new instance using that element should get the new implementation.

Yes, I came to the same realization. I think we don't disagree here? That's what I tried to model.

I agree it would be nice to "remount in place" but practically speaking I didn't want to get very "custom" in reconciliation unless we agreed that I'd need to.